tools/pythonpkg/duckdb_python.cpp

```diff
@@ -487,6 +487,35 @@ struct DuckDBPyResult {
 		return py::module::import("pandas").attr("DataFrame").attr("from_dict")(fetchnumpy());
 	}
 
+	py::object fetch_arrow_table() {
+		if (!result) {
+			throw runtime_error("result closed");
+		}
+
+		auto pyarrow_lib_module = py::module::import("pyarrow").attr("lib");
+
+		auto batch_import_func = pyarrow_lib_module.attr("RecordBatch").attr("_import_from_c");
+		auto from_batches_func = pyarrow_lib_module.attr("Table").attr("from_batches");
+		auto schema_import_func = pyarrow_lib_module.attr("Schema").attr("_import_from_c");
+		ArrowSchema schema;
+		result->ToArrowSchema(&schema);
+		auto schema_obj = schema_import_func((uint64_t)&schema);
+
+		py::list batches;
+		while (true) {
+			auto data_chunk = result->Fetch();
+			if (data_chunk->size() == 0) {
+				break;
+			}
+			ArrowArray data;
+			data_chunk->ToArrowArray(&data);
+			ArrowSchema schema;
+			result->ToArrowSchema(&schema);
+			batches.append(batch_import_func((uint64_t)&data, (uint64_t)&schema));
+		}
+		return from_batches_func(batches, schema_obj);
+	}
```

Comment:
As soon as pyarrow has built-in support for the C interface stream/iterator (i.e. hopefully in the next release!) this can be basically dropped